src/atldld/sync.py

```diff
@@ -519,6 +520,174 @@ def corners_sagittal(r, dataset_id):
     return section_number_corner_mode, image_id_corner_mode
 
 
+def get_transform_parallel(
+    slice_coordinate: float,
+    matrix_2d: np.ndarray,
+    matrix_3d: np.ndarray,
+    axis: str = "coronal",
+    ds_r: int = 1,
+    ds_i: int = 0,
+) -> DisplacementField:
+    """Compute displacement field between the reference space and the image.
+
+    Parameters
+    ----------
+    slice_coordainte : float
+        Value of the `axis` coordinate at which the image was sliced.
+    matrix_2d : np.array
+        Matrix of shape `(2, 3)` representing a 2D linear transformation.
+    matrix_3d : np.array
+        Matrix of shape `(3, 4)` representing a 3D linear transformation.
+    axis : str, {"coronal", "sagittal", "transverse"}
+        Axis along which the slice was made.
+    ds_r : int
+        Downscaling of the reference space grid. If set to 1 no
+        downsampling takes place. The higher the value the smaller the grid
+        in the reference space and the faster the matrix multiplication.
+    ds_i : int
+        The downloaded image will have both the height and the width
+        downsampled by `2 ** ds_i`.
+
+    Returns
+    -------
+    DisplacementField
+        Displacement field representing the transformation between the
+        reference space and the image. Note that one can directly use it
+        to register raw histological images to the reference space.
+    """
+    refspace = (  # order matters
+        ("coronal", 13200),
+        ("transverse", 8000),
+        ("sagittal", 11400),
+    )
+    axis_fixed = [i for i, a in enumerate(refspace) if a[0] == axis][0]
+    axes_variable = [i for i, a in enumerate(refspace) if a[0] != axis]
+
+    grid_shape = [refspace[i][1] // ds_r for i in axes_variable]
+    n_pixels = np.prod(grid_shape)
+
+    coords_ref = np.ones((4, n_pixels))
+    coords_ref[axis_fixed] *= slice_coordinate
+    coords_ref[axes_variable] = np.indices(grid_shape).reshape(2, -1) * ds_r
+
+    coords_temp = np.ones((3, n_pixels))
+    coords_temp[[0, 1]] = (matrix_3d @ coords_ref)[:2]  # (3, 4) x (4, n_pixels)
+
+    coords_img = matrix_2d @ coords_temp  # (2, 3) x (3, n_pixels)
+
+    tx = coords_img[0].reshape(grid_shape) / (2 ** ds_i)
+    ty = coords_img[1].reshape(grid_shape) / (2 ** ds_i)
+
+    df = DisplacementField.from_transform(tx, ty)  # `from_transform` not annotated
+
+    return df  # type: ignore
+
+
+def download_dataset_parallel(
+    dataset_id: int,
+    ds_r: int = 25,
+    detection_xy: Tuple[Union[int, float], Union[int, float]] = (0, 0),
+    include_expression: bool = False,
+    ds_i: int = 0,
+) -> Generator[
+    Union[
+        Tuple[int, float, np.ndarray, DisplacementField],
+        Tuple[int, float, np.ndarray, DisplacementField, np.ndarray],
+    ],
+    None,
+    None,
+]:
```

Comment:
- I think it's recommended to just annotate `float` instead of `Union[int, float]` because [`float` already implies `int`](https://www.python.org/dev/peps/pep-0484/#the-numeric-tower)
- That's a huge type annotation 😅 I guess there's nothing we can do about it at this point. Maybe some refactoring can be done in the future to avoid the ambiguous return type.

By the way for the cases like this where parameter/return types depend on each other there's the `overload` decorator in `typing`. It's probably an overkill to add overloads here. But just for info, it would look something like this:
```python
@overload
def download_dataset_parallel(
    dataset_id: int,
    ds_r: int = ...,
    detection_xy: Tuple[float, float] = ...,
    include_expression: Literal[False] = ...,
    ds_i: int = ...,
) -> Generator[
    Tuple[int, float, np.ndarray, DisplacementField],
    None,
    None,
]:
    ...


@overload
def download_dataset_parallel(
    dataset_id: int,
    ds_r: int = ...,
    detection_xy: Tuple[float, float] = ...,
    include_expression: Literal[True] = ...,
    ds_i: int = ...,
) -> Generator[
    Tuple[int, float, np.ndarray, DisplacementField, np.ndarray],
    None,
    None,
]:
    ...
```

Reply:
Interesting! Thank you! Something to keep in mind! Anyway, I fixed the `float` thing here 574521fc8d64d604e7469e00c534d48db5352f46